### PR TITLE
 automation: add additional headers to requestor job

### PR DESCRIPTION
**Pull request by njmulsqb**
This implementation currently only supports additional headers via YAML file, GUI implementation is ~~pending~~ done.
resolves zaproxy/zaproxy#6917

### CODE REVIEW

**Comment by ArkaprabhaChakraborty** [comment]
You need a spotlessApply for the tests to pass

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

**Comment by njmulsqb** [comment]
> You need a spotlessApply for the tests to pass

Can you please tell me what it is and when is it required so that I take care of it in future as well


> To address the DCO requirement you'll need to sign-off the commit(s):

Oh I thought they're signed-off, nevertheless I'll see this. Thanks

**Comment by ArkaprabhaChakraborty** [comment]
Just run this in your terminal, within the zap-extension directory of course :)
``./gradlew spotlessApply``

**Comment by ArkaprabhaChakraborty** [comment]
PS for squashing [here](https://twitter.com/FValkyrie_17/status/1557808806825951234?s=20&t=CUqkYUaDBus-lJmKYEH4_Q)'s a little one-trick-pony which you may use. :)

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java:22]: This should be just before the class declaration and discuss the class (not file).

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java:165]: If really necessary this should be javadoc on Job#getType()

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java:230]: I'm sure this was handy for you, but it doesn't seem generally relevant to all uses/users.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/JobUtils.java:252]: I think if these are really necessary they should probably be above the line they apply to. (Throughout)

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:74]: I think this is evident, it's in `RequestorJob` and it's getting `"parameters"`

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:79]: Can `o` simply be given a more meaningful name instead of the comment?

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:90]: Neither of these seem necessary

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:123]: You can rename `req` ➡️ `request` which addresses the first half of the comment.
Is the second part really that much of a leap based on `getData().addRequest(request)`? To me it seems evident request is being added to the job's data... but maybe I'm over simplifying in my head.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:133]:
```suggestion
    /**
     * This method sets the headers plus other required components of an HTTP message and then sends the HTTP request.
     * */
```

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:143]:
```suggestion
            /** Adding custom headers to the HTTP message */
```

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:186]: `sendAndReceive` seems clear.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (njmulsqb) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:123]: `request` variable is already there, so changing `req` to `request` is gonna require some work.

Comment (kingthorin) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java]: Well it could be `modifiedRequest` or `newRequest` I dunno, I'd have to read through again but there should be a reasonable way to make it more obvious.

Comment (njmulsqb) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java]: Agreed!

**Comment by kingthorin** [comment]
Overall the code seems fine to me. I'm sure others will also have feedback on the comments. But functionally the code seem reasonable to me. I'll try to checkout this branch and give it a try on the weekend.

**Comment by njmulsqb** [comment]
Here's how YAML file looked like that I used to test this functionality
```
---
env:
  contexts:
    - name: "Default Context"
      urls:
        - "https://njmulsqb.engineer"
      includePaths: []
      excludePaths: []
      authentication:
        parameters: {}
        verification:
          method: "response"
          pollFrequency: 60
          pollUnits: "requests"
      sessionManagement:
        method: "cookie"
        parameters: {}
  parameters:
    failOnError: true
    failOnWarning: false
    progressToStdout: true
  vars: {}
jobs:
  - parameters:
      user: ""
    requests:
      - url: "https://njmulsqb.engineer"
        name: ""
        method: ""
        data: ""
        headers:
          header1: "value1"
          header2: "value2"
        responseCode: 200
    name: "requestor"
    type: "requestor"

```

**Comment by njmulsqb** [comment]
> Seems to need a spotlessApply again.

I ran it on zap-extensions only, is it required on zaproxy code too?

**Comment by kingthorin** [comment]
Nope you're good, it did only need it on zap-extensions. It's now failing because some of the unit tests are failing.

**Comment by njmulsqb** [comment]
Actually, I was not checking if customHeaders variable is null or not, so when the unit test was not giving customHeaders. the exception got threw. I have placed a check which is handling this usecase now.

**Comment by njmulsqb** [comment]
So, I am onto GUI implementation part now. One thing is when I add the following line to AddRequestDialog.java
```
   this.addTextField(HEADERS_PARAM, rule.getHeaders());
```
This gives error that the field expected string whereas `getHeaders()` is returning Map so running `.toString()` is what the IDE has suggested but this also fails because in the beginning when we open Add Request dialog there's nothing returned by all the getters as Request object is empty so running `.toString()` on a null value throws exception obviously.
Any ideas on how a Map should be dealt with in a textField? Also how are we going to give the user option to define headers? Does comma-separated strings sounds good?

Also, I have added line 
```
automation.dialog.addreq.customheaders = Custom Headers:
```
to messages.properties file but this is not turning into orange like others 
image

Does it require anything else to use strings from messages.properties?

**Comment by kingthorin** [comment]
I’d suggest using a two column table and allow the user to type/add rows.

**Comment by njmulsqb** [comment]
> I’d suggest using a two column table and allow the user to type/add rows.

Like this?
| Header    | Value     |   
|-----------|-----------|
| X-Header1 | dsfsasgs  |
| X-Header2 | dsgsdgfdg |

**Comment by kingthorin** [comment]
Yup

**Comment by njmulsqb** [comment]
Okay. Any idea about showing map in textField and messages.properties issue I mentioned?

**Comment by kingthorin** [comment]
That was kinda my point use a table not a text field.

**Comment by njmulsqb** [comment]
ah, got it now. Thanks mate!

**Comment by njmulsqb** [comment]
This is the code of somewhat buggy GUI implementation. Here's how it looks like 
image

Currently, I am struggling with following issues, if you know what's wrong kindly let me know I have spent some time on it but stuck
1. After adding custom headers to the request if you save/close the dialog and reopen it, you wont see the custom headers table populated again. Re-opening the dialog clears the previously stored custom headers from the rule.
2. Modify button doesn't gets enabled even after I have selected some row.
3. Removing and re-adding a custom header doesn't closes the dialog box even after the new entry is added
4. Even selecting the row throws null exception
5. Removing a header does not immediately remove the row from table, it only does when you click on the removed entry once, then it disappears.
I know its very buggy and not really usable at the moment, I have just copied/pasted the code from similar implementations and modified it to use custom Headers. I am pretty new to GUI so I'm sure I must have made some silly mistakes and there would be some more bugs in there as well. 
If any of the above mentioned bug's issue is something you know is trivial to fix so kindly let me know.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:166]: this has to be moved in the constructor I believe...

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:183]: same here

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:199]: same

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:208]: probably the same for this too ...

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:105]: I'd probably use a Map to "hold" the buttons and add the actionListeners :).

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:152]: Should be I18Ned

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:164]: Should be I18Ned

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:180]: Same ... I18N

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/gui/AddRequestDialog.java:186]: I18N

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:83]: probably it should be List<?>.. rather than ? you should specify the type that's expected maybe? :)

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java
Comment (ArkaprabhaChakraborty) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:151]: Same ... left hand side types can be specified I guess :)

**Review by ArkaprabhaChakraborty — COMMENTED** [review]
Couple of things I found... might fix your problems

**Comment by njmulsqb** [comment]
Hi,

I have fixed the bugs and the functionality is now working fine now. Please review the code, the PR is ready to be merged :)

**Comment by thc202** [comment]
The build is failing (seems to need a Spotless apply).

The changes should have unit tests.

**Comment by njmulsqb** [comment]
Can you please elaborate what sort of unit tests are required?

**Comment by thc202** [comment]
They are in [`RequestorJobUnitTest`](https://github.com/zaproxy/zap-extensions/blob/8efae792f0e121b503f4da2913733bbdedd2c4ee/addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java), should verify the messages sent have the expected headers.

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java
Comment (kingthorin) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:372]: This is a good start.

I'd also suggest adding a test that has:
- a `requests:` and `headers:` section but no actual headers
- a header name (key) but no value
- a header name (key) and empty string (`""`)

**Review by psiinon — CHANGES_REQUESTED** [review]
The GUI works well 😁 
Just need an update to the help page: https://github.com/zaproxy/zap-extensions/blob/main/addOns/automation/src/main/javahelp/org/zaproxy/addon/automation/resources/help/contents/job-requestor.html

**Comment by psiinon** [comment]
Can you also update the templates:

- [requestor-min.yaml](https://github.com/zaproxy/zap-extensions/blob/main/addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/requestor-min.yaml)
- [requestor-max.yaml](https://github.com/zaproxy/zap-extensions/blob/main/addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/requestor-max.yaml)

It would be worth double checking the job works fine with an empty set of headers 😉

**Comment by thc202** [comment]
The branch needs to be rebased.

**Comment by njmulsqb** [comment]
> The branch needs to be rebased.

You meant merging the 2 commits?

**Comment by njmulsqb** [comment]
Hi, to rebase the `feature-custom-headers` branch, I have performed `git fetch --all` on my local main branch but now its saying 
```
Switched to branch 'main'
Your branch is ahead of 'origin/main' by 265 commits.
  (use "git push" to publish your local commits)
```
I don't know what these 265 commits are about, I haven't made any changes in the main branch. I have done some googling but cant figure out.
Any clues?

**Comment by njmulsqb** [comment]
Running these two 
    git fetch upstream
    git rebase upstream/main
have produced some conflicts. I am tangled, please help :P

**Comment by thc202** [comment]
What is the conflict?

**Comment by thc202** [comment]
> Any clues?

It's just saying that your remote origin is missing those 256 commits, since you updated your local main branch. You can push the main branch if you want to include those changes in your remote.

**Comment by njmulsqb** [comment]
> > Any clues?
> 
> It's just saying that your remote origin is missing those 265 commits, since you updated your local main branch. You can push the main branch if you want to include those changes in your remote.

Means that changes have been pulled from zaproxy/zap-extensions locally which are not available in njmulsqb/zap-extensions? But the github web says otherwise 
image

**Comment by njmulsqb** [comment]
> What is the conflict?

Merge conflicts in some files like this one `api.html`. I have not worked on any of these files.
image

**Comment by kingthorin** [comment]
> Means that changes have been pulled from zaproxy/zap-extensions locally which are not available in njmulsqb/zap-extensions? But the github web says otherwise

Because you didn’t fetch origin so it’s comparing against our dated info.

**Comment by thc202** [comment]
> I have not worked on any of these files.

Abort the rebase and tell what steps are you doing.

**Comment by njmulsqb** [comment]
Aborted.

It still says that I have my local `main` is ahead of origin/main by 265 commits.

By the way, this PR is on a separate branch `feature-custom-headers` if by any means I have messed up the local `main` can we rebase `feature-custom-headers` directly from the remote origin i.e zaproxy/zap-extensions instead of local main

**Comment by thc202** [comment]
You should be rebasing on `upstream/main`.

**Comment by njmulsqb** [comment]
Ok so I have ran 
1. `git fetch upstream`
2. `git rebase upstream/main`


on `feature-custom-headers`. It says successfully rebased, but git status is saying this now
```
C:\Users\najamul.saqib\Documents\GitHub\zap-extensions>git status
On branch feature-custom-headers
Your branch and 'origin/feature-custom-headers' have diverged,
and have 593 and 2 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

nothing to commit, working tree clean

```
What should be my next step? Should I take pull and then force push on this PR?

**Comment by thc202** [comment]
That's the expected state, just force push to your remote.

**Comment by njmulsqb** [comment]
Are we good now?

**Comment by thc202** [comment]
Yes, thank you.

**Comment by kingthorin** [comment]
If you diff the files are they the same? Are your line endings correct?

File: addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/requestor-max.yaml
Comment (thc202) [inline addOns/automation/src/main/resources/org/zaproxy/addon/automation/resources/requestor-max.yaml:11]: IMO the header should not be added, by default it shouldn't include any.

**Comment by njmulsqb** [comment]
Hi, a bit off-topic but will this PR be counted towards my hacktoberfest contribution? I created this before October but it will be merged during October.

**Comment by psiinon** [comment]
Fine by me 😁

**Pull request closed without merging**
